flex strict mode

jonapich opened the pull request:
When strict=True and the payload can't be deserialized into the hint, raise instead of returning the value untouched.

dotboris commented on coveo-functools/coveo_functools/exceptions.py:
```diff
@@ -8,3 +11,13 @@ class Flexception(CoveoFunctoolsException):  # sorry for the pun :socanadian:
+    def __init__(self, value: Any, hint: Any, contains: Any) -> None:
+        super().__init__(
+            f"I don't know how to fit {type(value)} "
+            f"into {hint}{' of ' + str(contains) if contains else ''}"
```
I'm not sure if it's a good idea but you might want to add the value to the message or add the value as a separate field. When someone sees this error, they'll need to dig into the code and start debugging to figure out what the problematic value was. If it's in the message then it's easy to tell what's going on.

I'm not sure if this is a good idea because these values may contain sensitive information or they may be gigantic.

I think that if you add the value as a field then platforms like Sentry can interface it in their UI (I think). They have a layer that removes secrets which should work on that.

jonapich replied:
Good call about the log fields, I'll add that. I had the exact same concern about sensitive or gigantic information.

dotboris commented on coveo-functools/coveo_functools/flex/deserializer.py:
```diff
@@ -61,23 +62,25 @@ def convert_kwargs_for_unpacking(dirty_kwargs: Dict[str, Any], *, hint: MetaHint
 
 
-def deserialize(value: Any, *, hint: Union[T, Type[T]]) -> T:
+def deserialize(value: Any, *, hint: Union[T, Type[T]], strict: bool = False) -> T:
```
Maybe this could be called `raise` or something similar that's not a reserved keyword (python might not like you using `raise`). This would make the usage pretty obvious to people who have never seen the option or are not familiar with how `flex` outputs values.

dotboris reviewed (commented):
PR looks good 👍 

I have a couple of ideas to potentially improve the UX/DX of this tho they're non-blocking.

jonapich commented:
@dotboris @marier-nico I changed this PR considerably:

1. Instead of `strict` as a bool, I used a system similar to `str.encode()` where you can write e.g. `encode(errors='ignore')`.
2. I added a deprecation warning around the default value to encourage users to either specify raise or ignore. I think the default should become `raise` at some point, but I don't want to introduce breaking changes.

klalumiere commented on coveo-functools/tests_functools/test_flex_deserializer.py:
```diff
@@ -314,3 +317,62 @@ def test_return() -> str:  # noqa
+) -> None:
+    """The legacy behavior would not handle the TypeError that occurs
+    when the value's type is correct but its content is wrong."""
+    with caplog.at_level(logging.ERROR), pytest.raises(TypeError):
```
I don't understand why you capture the log here since you don't look at it afterwards. :thinking:

klalumiere commented on coveo-functools/coveo_functools/flex/deserializer.py:
```diff
@@ -55,32 +61,91 @@ def convert_kwargs_for_unpacking(dirty_kwargs: Dict[str, Any], *, hint: MetaHint
+
+    example #1: `range` requires the `stop` argument, which is missing:
+
+        >>> deserialize({}, hint=range, errors='raise')  # doctest: +IGNORE_EXCEPTION_DETAIL
+        Traceback (most recent call last):
+        TypeError: range expected 1 argument, got 0
+        >>> deserialize({}, hint=range, errors='deprecated')  # doctest: +IGNORE_EXCEPTION_DETAIL
+        Traceback (most recent call last):
+        TypeError: range expected 1 argument, got 0
```
I think there might be a copy-paste error here. Both cases look very similar.

jonapich replied:
the first one has raise, the second has deprecated. The behavior is the same. However in the other example, the behavior of "deprecated" is different. That's what I tried to illustrate here. The middle choice (deprecated) is not consistent, since it would raise in one situation and ignore in the other.

klalumiere commented on coveo-functools/coveo_functools/flex/deserializer.py:
```diff
@@ -1,4 +1,6 @@
 import inspect
+import logging
+import warnings
```
TIL :heart: :nerd_face:

klalumiere commented on coveo-functools/coveo_functools/flex/deserializer.py:
```diff
@@ -55,32 +61,91 @@ def convert_kwargs_for_unpacking(dirty_kwargs: Dict[str, Any], *, hint: MetaHint
+        >>> deserialize([], hint=dict, errors='ignore')
+        []
     """
+    valid_error_modes = "raise", "ignore", "silent", "deprecated"
+    if errors not in valid_error_modes:
+        raise ValueError(f"'{errors=}' is not valid, {valid_error_modes=}")
+
+    if errors == "deprecated":
+        warnings.warn(
+            "Please specify the error behavior when calling `flex.deserialize`. "
+            "Recommended fix: specify `errors='raise'` to catch deserialization errors. "
```
Another choice would have been to use something like an `enum` to represent the error mode. The trade-off I see are

**Good points of using `enum`**:
1. No typos are possibles (more precisely, typos would be caught at static analysis time by e.g. `mypy`).

**Good points of using a _string_**:
1. People need to import the `enum` to use it. Hence, the `Recommend fix` won't fit on a single line.

I think in the current context, I'd value more the `enum` than the string. However, you already implemented things with a string, and I'm not sure the benefit of using an `enum` is worth the time required to do the change. :thinking:

jonapich replied:
I'm a BIG enum fan, but I don't want people to be forced to import 2 symbols from flex. 

For your information, I used the `Literal` type, so it's actually type safe! :)

klalumiere replied:
I missed that! :heart:
